---
title: "Alpha diversity analyses for published article"
subtitle: "Metagenomic characterization of the supragingival plaque and salivary microbiotas during initiation and resolution of experimental gingivitis"
author: "Caroline Schou Nielsen"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document: 
    toc: yes
    keep_md: no
---

```{r setup}
rm(list = ls())
gc()
# invisible(lapply(paste0('package:', names(sessionInfo()$otherPkgs)), detach, character.only=TRUE, unload=TRUE))
# options(java.parameters = "-Xmx80000m")
```

```{r setup2}
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
##knitr::opts_chunk$set(cache=TRUE)
##knitr::opts_chunk$set(echo = TRUE)
##knitr::opts_chunk$set(dev = "ragg_png") # https://ragg.r-lib.org/index.html
##knitr::opts_chunk$set(fig.width = 6) #https://r4ds.had.co.nz/graphics-for-communication.html#figure-sizing
knitr::opts_chunk$set(fig.asp = 0.618)
knitr::opts_chunk$set(fig.show = "hold")
knitr::opts_chunk$set(fig.show = "70%")
knitr::opts_chunk$set(fig.align = "center")
```

```{r setup3}
library(tidyverse)
library(readxl)
library(readr)
library(speedyseq)
library(microViz)
library(microbiome)
library(here)
```

```{r setup4}
# Import functions
'%!in%' <- function(x,y)!('%in%'(x,y))

source_dir = "https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/"

source(paste0(source_dir,"phyloseq_normalisation.R"))
source(paste0(source_dir,"phyloseq_alpha.R"))
source(paste0(source_dir,"phyloseq_varia.R"))
```

```{r setup5}
# Alpha diversity parameters
source("https://raw.githubusercontent.com/fconstancias/KU_Caroline/refs/heads/main/code/functions/phyloseq_functions.R")

# compute_plot_alpha()
# details:

#' @title Compute Alpha Diversity Plot and Statistical Summaries
#' @description
#' This function generates plots and statistical summaries for alpha diversity measures 
#' over time across different sample groups. The function returns a list containing summary
#' statistics, plots, and results from statistical tests.
#' @author Florentin Constancias
#' @param alpha_long_df Data frame in long format containing alpha diversity values. 
#'        Should have columns specified by `alpha_measures`, `x`, `y`, `color_point`, etc.
#' @param alpha_measures Character vector of alpha diversity measures to analyze and plot.
#' @param pd Position dodge to adjust plot element spacing (default is 0.3).
#' @param x Character string for x-axis variable (e.g., "Time").
#' @param y Character string for y-axis variable (e.g., "value").
#' @param color_point Column name for coloring points (e.g., "Subject").
#' @param shape_point Column name for point shape (e.g., "Sample").
#' @param group_point Column name for grouping points (e.g., "interaction(Sample,Subject)").
#' @param group_boxplot Column name for grouping boxplot elements (e.g., "interaction(Sample,Time)").
#' @param group_line Column name for grouping line elements (e.g., "Subject").
#' @param facet_formula Formula for faceting plots (default is "alphadiversiy ~ Sample + cluster_Dtp2").
#' @param col_pal Color palette for points and lines.
#' @param fill_pal Fill palette for boxplots.
#' @param stat_formula Formula for statistical testing (default is "value ~ Time").
#' @param anova_test_formula Formula for ANOVA testing (default is "value ~ Time*Sample + Error(Subject/(Time*Sample))").
#' @param group_by_stats Variables for grouping in statistical comparisons.
#' @param padjust_method Method for p-value adjustment (default is "fdr").
#' @param stat_paired Logical; whether statistical tests are paired (default is FALSE).
#' @param ref_group_stat Reference group for pairwise statistical tests.
#' @return A list with summary data frames, ggplot objects, and statistical test results.
```

```{r MetaPhlAn}
# Load phyloseq MetaPhlAn object
# Not included in GitHub repo due to size
# See README for data access information

load(here::here("01_data.Rdata"))

# Quick control that data is right
ps_up %>% 
  sample_data() %>% 
  data.frame() %>% 
  DT::datatable()
```

```{r HUMAnN}
# Load phyloseq HUMAnN object
# Not included in GitHub repo due to size
# See README for data access information

load(here::here("01_humann_data.Rdata"))

# Quick control that data is right
pw_mia %>% 
  sample_data() %>% 
  data.frame() %>% 
  DT::datatable()
```

```{r plot_positions}
# Global ggplot position settings (used across figures)
pd <- position_dodge(0.3)
```

```{r MetaPhlAn data transformation}
# MetaPhlAn data transformation for alpha diversity analysis
alphas <- ps_up %>% 
  subset_taxa(Class != "UNCLASSIFIED") %>%
  transform_sample_counts(function(x) x/sum(x) * 1) %>% 
  phyloseq_alphas()
```

```{r select alpha diversity measures}
# Select alpha diversity measures of interest 
alpha_measures <- c("diversity_shannon") 
# Other available measures:
# "observed", "diversity_coverage", "diversity_inverse_simpson", "evenness_pielou"

alpha_long_df <- alphas %>% 
  pivot_longer(
    cols = all_of(alpha_measures),
    names_to = "alphadiversity",
    values_to = "value",
    values_drop_na = TRUE
  ) %>%
  mutate(
    alphadiversity = fct_relevel(alphadiversity, alpha_measures)
  )

alpha_long_df
```

```{r MetaPhlAn boxplot}
# Boxplot of MetaPhlAn (taxonomy) alpha diversity
tax_alpha <- alpha_long_df %>% 
   compute_plot_alpha(facet_formula = "alphadiversiy ~ Sample", 
                     group_by_stats = c("alphadiversiy", "Sample"), 
                     stat_formula = "value ~ Time",
                     padjust_method = "fdr")
                     
tax_alpha$alpha_plot + 
  ggpubr::rotate_x_text(60)
```

```{r MetaPhlAn statistics}
# Wilcoxon pairwise tests for MetaPhlAn (taxonomy) alpha diversity comparisons
tax_alpha$wilcox.test_stat %>% 
 DT::datatable()
```

```{r HUMAnN data transformation}
# HUMAnN data transformation for alpha diversity analysis
func_alphas <- pw_mia %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>% 
  phyloseq_alphas()
```

```{r select alpha diversity measures}
# Select alpha diversity measures of interest 
alpha_measures <- c("diversity_shannon") 
# Other available measures:
# "observed", "diversity_coverage", "diversity_inverse_simpson", "evenness_pielou"

func_alpha_long_df <- func_alphas %>% 
  pivot_longer(
    cols = all_of(alpha_measures),
    names_to = "alphadiversity",
    values_to = "value",
    values_drop_na = TRUE
  ) %>%
  mutate(
    alphadiversity = fct_relevel(alphadiversity, alpha_measures)
  )

func_alpha_long_df
```

```{r HUMAnN boxplot}
# Boxplot of HUMAnN (functional) alpha diversity
func_alpha <- func_alpha_long_df %>% 
   compute_plot_alpha(facet_formula = "alphadiversiy ~ Sample", 
                     group_by_stats = c("alphadiversiy", "Sample"), 
                     stat_formula = "value ~ Time",
                     padjust_method = "fdr")
                     
func_alpha$alpha_plot + 
  ggpubr::rotate_x_text(60)
```

```{r HUMAnN statistics}
# Wilcoxon pairwise tests for HUMAnN (functional) alpha diversity comparisons
func_alpha$wilcox_test %>% 
  DT::datatable()
```

```{r session info}
sessionInfo()
```
